### PR TITLE
AMBA AXI4-Lite master

### LINKED ISSUE #434: Implement AXI-lite master
We use currently a wishbone bus, which is pretty easy to implement, but has limited throughput. To increase it we can switch to AXI-lite, so please implement a master to such bus. If possible take care about interface compatibility with wishbone masters.

### PR DESCRIPTION
**Pull request by Durchbruchswagen**
Implementation of AMBA AXI4-Lite master

Resolves #434

### CODE REVIEW

File: coreblocks/peripherals/axi_lite.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+            ("write_address", write_address),
+            ("write_data", write_data),
+            ("write_response", write_response),
+            ("read_address", read_address),
+            ("read_data", read_data),
+        ]
+
+
+@unique
+class States(Enum):
```
Comment (tilk) [inline coreblocks/peripherals/axi_lite.py:87]: This is a single state, not a set of states.
```suggestion
class State(Enum):
```

File: coreblocks/peripherals/axi_lite.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
+from amaranth import *
+from amaranth.hdl.rec import DIR_FANIN, DIR_FANOUT
+from transactron import Method, def_method, TModule
+from transactron.core import Transaction
+from transactron.lib.connectors import Forwarder
+from enum import Enum, unique
```
Comment (tilk) [inline coreblocks/peripherals/axi_lite.py:6]: Shouldn't you use an Amaranth enum?

File: coreblocks/peripherals/axi_lite.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+            ("data", axil_params.data_width),
+            ("strb", axil_params.data_width // 8),
+        ]
+
+        self.wr_response_layout = [
+            ("resp", 2),
+        ]
+
+    def state_machine_request(self, state, prot, address, channel, forwarder, method, m):
+        with m.If(state == States.IDLE):
```
Comment (tilk) [inline coreblocks/peripherals/axi_lite.py:189]: Why not use Amaranth's FSMs?

File: coreblocks/peripherals/axi_lite.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+            m.d.sync += channel.val.eq(0)
+            with m.If(method.run):
```
[the remark this answers is not in the document]
Comment (tilk) [reply coreblocks/peripherals/axi_lite.py:191]: Actually, is the forwarder even doing something useful? Why didn't you follow the pattern from the Wishbone implementation?

File: coreblocks/peripherals/axi_lite.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+                            m.d.sync += channel.data.eq(rtval.data)
+                        if prot:
+                            m.d.sync += channel.prot.eq(rtval.prot)
+                        else:
+                            m.d.sync += channel.strb.eq(rtval.strb)
+                with m.Else():
+                    m.d.sync += channel.val.eq(0)
+                    m.d.sync += state.eq(States.IDLE)
+
+    def result_handler(self, data, channel, forwarder, m):
```
Comment (tilk) [inline coreblocks/peripherals/axi_lite.py:222]: Again, the arguments problem.

File: test/peripherals/test_axi_lite.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+            return m
+
+    def test_manual(self):
+        almt = TestAXILiteMaster.AXILiteMasterTestModule()
+
+        def master_process():
+            # read request
+            yield from almt.read_address_request_adapter.call(addr=5, prot=0)
+
+            yield
```
Comment (tilk) [inline test/peripherals/test_axi_lite.py:95]: Why is there a delay between each request? This isn't testing the maximum throughput scenario.

File: test/peripherals/test_axi_lite.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        yield self.axi_lite.write_response.resp.eq(resp)
+        yield self.axi_lite.write_response.val.eq(1)
+        yield
+        yield self.axi_lite.write_response.val.eq(0)
+
+
+class TestAXILiteMaster(TestCaseWithSimulator):
+    class AXILiteMasterTestModule(Elaboratable):
+        def __init__(self):
+            pass
```
Comment (tilk) [inline test/peripherals/test_axi_lite.py:75]: Is this needed?

File: test/peripherals/test_axi_lite.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+            yield from slave.slave_ra_wait()
+            yield from slave.slave_ra_ready(1)
+            yield from slave.slave_ra_verify(5, 0)
+            yield
+            yield Settle()
+
+            yield from slave.slave_ra_wait()
+            yield from slave.slave_rd_wait()
+            yield from slave.slave_ra_ready(1)
```
Comment (tilk) [inline test/peripherals/test_axi_lite.py:118]: `rdy` is already 1 here, why do you do that?

File: test/peripherals/test_axi_lite.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+    def slave_wr_respond(self, resp=0):
+        assert (yield self.axi_lite.write_response.rdy)
+        yield self.axi_lite.write_response.resp.eq(resp)
+        yield self.axi_lite.write_response.val.eq(1)
+        yield
+        yield self.axi_lite.write_response.val.eq(0)
+
+
+class TestAXILiteMaster(TestCaseWithSimulator):
```
Comment (tilk) [inline test/peripherals/test_axi_lite.py:72]: Did you look at the waveforms for the test? Did you test for all sensible scenarios, including maximum throughput, master delays and slave delays?

A randomized test would be best, but a comprehensive manual test will be acceptable for now. (The test for the Wishbone master is not great - it's one of the first written.)

File: test/peripherals/test_axi_lite.py (hunk 7 above)
Comment (lekcyjna123) [reply test/peripherals/test_axi_lite.py:75]: You can use `SimpleTestCircuit`
https://github.com/kuznia-rdzeni/coreblocks/blob/e898af5394c8a436ac9f711d7c87a3cc3245a151/test/common/infrastructure.py#L34C7-L34C24
which will wrap all methods with TestbenchIOs.

File: coreblocks/peripherals/axi_lite.py (hunk 1 above)
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: After changing `state_machine_request` to use Amaranth's FSM this class is obsolete and was removed

File: coreblocks/peripherals/axi_lite.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: After changing state_machine_request to use Amaranth's FSM this became obsolete and was removed

File: coreblocks/peripherals/axi_lite.py [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+        self.wd_forwarder_layout = [
+            ("data", axil_params.data_width),
+            ("strb", axil_params.data_width // 8),
+        ]
+
+        self.wr_response_layout = [
+            ("resp", 2),
+        ]
+
+    def state_machine_request(self, state, prot, address, channel, forwarder, method, m):
```
[the remark this answers is not in the document]
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: Done, also after refactor I managed to reduce numbers of arguments

File: coreblocks/peripherals/axi_lite.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: After refactor the forwarders for requests were remove (they were not doing anything useful, their functionality was replaced by a single signal for each request). The `m.If(method.run)` now only changes state `Idle` to `Active`

File: coreblocks/peripherals/axi_lite.py [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+    """
+
+    def __init__(self, axil_params: AXILiteParameters):
+        self.axil_params = axil_params
+        self.axil_layout = AXILiteLayout(self.axil_params).axil_layout
+        self.axil_master = Record(self.axil_layout)
+
+        self.FSM_ra = Signal(States, reset=States.IDLE)
+        self.FSM_wa = Signal(States, reset=States.IDLE)
+        self.FSM_wd = Signal(States, reset=States.IDLE)
```
[the remark this answers is not in the document]
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: No particular reason beside my negligence. After changing `state_machine_request` to use Amaranth's FSM those signals became obsolete and were removed

File: coreblocks/peripherals/axi_lite.py (hunk 5 above)
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: Done the same thing that I did with `state_machine_request`

File: test/peripherals/test_axi_lite.py (hunk 6 above)
Comment (Durchbruchswagen) [reply test/peripherals/test_axi_lite.py]: It was the result of me not fully understanding how tests works, but after looking at waveforms I saw that this yield was unnecessary and I removed it

File: test/peripherals/test_axi_lite.py (hunk 8 above)
Comment (Durchbruchswagen) [reply test/peripherals/test_axi_lite.py]: copy-paste mistake. I removed it.

File: test/peripherals/test_axi_lite.py (hunk 9 above)
Comment (Durchbruchswagen) [reply test/peripherals/test_axi_lite.py:72]: I added more tests cases. It now include following scenarios:

- 4 read requests in a row (testing max throughput)
- 4 write requests in a row (testing max throughput)
- master issue requests and waits for `ready` signal from slave
- slave returns response and waits for 'ready' signal from master

I also looked at waveforms after refactoring the test and signals on all channels looks nice

File: coreblocks/peripherals/axi_lite.py [hunk 12]
```diff
@@ -0,0 +1,283 @@
+from amaranth import *
+from amaranth.hdl.rec import DIR_FANIN, DIR_FANOUT
+from transactron import Method, def_method, TModule
+from transactron.core import Transaction
+from transactron.lib.connectors import Forwarder
+from enum import Enum, unique
+
```
[the remark this answers is not in the document]
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py:8]: Done, added `__all__` with `AXILiteParameters` and `AXILiteMaster`

File: coreblocks/peripherals/axi_lite.py [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+        Whether the layout should be generated for master side
+        (if false it's generatd for the slave side)
+
+    Attributes
+    ----------
+    axil_layout: Record
+        Record of a AXI-Lite bus.
+    """
+
+    def __init__(self, axil_params, *, master=True):
```
[the remark this answers is not in the document]
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: Done, added type annotations

File: coreblocks/peripherals/axi_lite.py [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+
+    Parameters
+    ----------
+    axil_params: AXILiteParameters
+        Parameters for bus generation/
+
+    Attributes
+    ----------
+    axil_master: Record(like AXILiteLayout)
+        AXI-Lite bust output
```
[the remark this answers is not in the document]
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: Done

File: coreblocks/peripherals/axi_lite.py [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+    ACTIVE = 1
+
+
+class AXILiteMaster(Elaboratable):
+    """AXI-Lite master interface.
+
+    Parameters
+    ----------
+    axil_params: AXILiteParameters
+        Parameters for bus generation/
```
[the remark this answers is not in the document]
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: Done

File: coreblocks/peripherals/axi_lite.py [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+
+    Attributes
+    ----------
+    axil_master: Record(like AXILiteLayout)
```
[the remark this answers is not in the document]
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: No, I don't. Moved `axil_master` to `elaborate`

File: coreblocks/peripherals/axi_lite.py [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+    def __init__(self, axil_params: AXILiteParameters):
+        self.axil_params = axil_params
+        self.axil_layout = AXILiteLayout(self.axil_params).axil_layout
+        self.axil_master = Record(self.axil_layout)
+
+        self.FSM_ra = Signal(States, reset=States.IDLE)
+        self.FSM_wa = Signal(States, reset=States.IDLE)
+        self.FSM_wd = Signal(States, reset=States.IDLE)
+
+        self.generate_layouts(axil_params)
```
[the remark this answers is not in the document]
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: Done, created class `AXILiteMasterMethodLayouts` that generates all layouts for methods

File: coreblocks/peripherals/axi_lite.py [hunk 18]
```diff
@@ -0,0 +1,283 @@
...
+        @def_method(m, self.wd_request, ready=wd_forwarder.write.ready)
+        def _(arg):
+            wd_forwarder.write(m, data=arg.data, strb=arg.strb)
+
+        # write_response
+        self.result_handler(False, self.axil_master.write_response, wr_forwarder, m)
+
+        @def_method(m, self.wr_response)
+        def _():
+            return wr_forwarder.read(m)
```
[the remark this answers is not in the document]
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py:288]: I managed to remove forwarders from all request methods. If we want to remove forwarders from responses I will need more time to think how to remove them while keeping max throughput

File: coreblocks/peripherals/axi_lite.py [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+
+    Attributes
+    ----------
+    axil_layout: Record
+        Record of a AXI-Lite bus.
+    """
+
+    def __init__(self, axil_params: AXILiteParameters, *, master: bool = True):
+        write_address = [
+            ("val", 1, DIR_FANOUT if master else DIR_FANIN),
```
Comment (tilk) [inline coreblocks/peripherals/axi_lite.py:45]: Does `val` mean `valid`? If so, use `valid`, as that's how this signal is named in the spec, and `val` is typically used as a short for `value`.

File: coreblocks/peripherals/axi_lite.py [hunk 20]
```diff
@@ -0,0 +1,290 @@
...
+        if is_address_channel:
+            m.d.sync += channel.addr.eq(arg.addr)
+            m.d.sync += channel.prot.eq(arg.prot)
+        else:
+            m.d.sync += channel.data.eq(arg.data)
+            m.d.sync += channel.strb.eq(arg.strb)
+        m.d.sync += channel.val.eq(1)
+
+    def state_machine_request(
+        self, m: TModule, method: Method, *, is_address_channel: bool, channel: Record, request_signal: Signal
```
Comment (lekcyjna123) [inline coreblocks/peripherals/axi_lite.py:196]: Is `is_address_channel` needed?

**Review by lekcyjna123 — APPROVED** [review]

File: coreblocks/peripherals/axi_lite.py (hunk 19 above)
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: Done, changed `val` to `valid`

File: coreblocks/peripherals/axi_lite.py [hunk 21]
```diff
@@ -0,0 +1,290 @@
...
+    """AXI-Lite master layouts for methods
+
+    Parameters
+    ----------
+    axil_params: AXILiteParameters
+        Patameters used to generate AXI-Lite master layouts
+
+    Attributes
+    ----------
+    ra_request_layout: Record
```
[the remark this answers is not in the document]
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: Done

File: coreblocks/peripherals/axi_lite.py (hunk 20 above)
Comment (Durchbruchswagen) [reply coreblocks/peripherals/axi_lite.py]: No it is not needed, I forgot to remove it after refactoring `state_machine_request`

**Review by tilk — APPROVED** [review]

**Pull request merged**